src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorProjectChangePublisher.cs

```diff
@@ -0,0 +1,233 @@
+{ex}");
+                }
+            }
+        }
+
+        protected virtual void SerializeToFile(ProjectSnapshot projectSnapshot, string publishFilePath)
+        {
+            var fileInfo = new FileInfo(publishFilePath);
+            using var writer = fileInfo.CreateText();
+            _serializer.Serialize(writer, projectSnapshot);
+        }
+
+        protected bool IsOnMainThread()
+        {
+            return _joinableTaskContext.IsOnMainThread;
+        }
+
+        internal void ProjectSnapshotManager_Changed(object sender, ProjectChangeEventArgs args)
+        {
+            if (!IsLSPEditorAvailable(args.ProjectFilePath))
+            {
+                return;
+            }
+
+            switch (args.Kind)
+            {
+                case ProjectChangeKind.DocumentRemoved:
+                case ProjectChangeKind.DocumentAdded:
+                case ProjectChangeKind.DocumentChanged:
+                case ProjectChangeKind.ProjectChanged:
+                    // These changes can come in bursts so we don't want to overload the publishing system. Therefore,
+                    // we enqueue publishes and then publish the latest project after a delay.
+
+                    EnqueuePublish(args.Newer);
+                    break;
+                case ProjectChangeKind.ProjectAdded:
+                    Publish(args.Newer);
+                    break;
+                case ProjectChangeKind.ProjectRemoved:
+                    RemovePublishingData(args.Older);
+                    break;
+            }
+        }
+
+        // Internal for testing
+        internal void EnqueuePublish(ProjectSnapshot projectSnapshot)
+        {
+            // A race is not possible here because we use the main thread to synchronize the updates
+            // by capturing the sync context.
+            _pendingProjectPublishes[projectSnapshot.FilePath] = projectSnapshot;
+
+            if (!_deferredPublishTasks.TryGetValue(projectSnapshot.FilePath, out var update) || update.IsCompleted)
+            {
+                _deferredPublishTasks[projectSnapshot.FilePath] = PublishAfterDelayAsync(projectSnapshot.FilePath);
+            }
+        }
+
+        // Internal for testing
+        // This needs to be overridable because if we try to get the LSPEditorFeatureDetector service
+        // in unit tests everything will break. We can't LSPEditorFeatureDetector pass to the constructor because the objects
+        // required to construct one are not available when DefaultRazorProjectChangePublisher is constructed, but
+        // they ARE available later, when this is called.
+        internal virtual bool IsLSPEditorAvailable(string projectFilePath)
+        {
+            if (_lspEditorFeatureDetector is null)
+            {
+                var componentModel = (IComponentModel)Shell.Package.GetGlobalService(typeof(SComponentModel));
+                _lspEditorFeatureDetector = componentModel.GetService<LSPEditorFeatureDetector>();
```

Comment:
I could have sworn I already answered this one, but apparently I forgot to hit send.

If I add `LSPEditorFeatureDetector` as a parameter to the constructor then the extension hangs when trying to construct a `LSPEditorFeatureDetector`, seemingly because all of the types required are not yet constructable. Given that I searched for what other usages were doing and [found this instance](https://github.com/dotnet/aspnetcore-tooling/blob/master/src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorEditorFactory.cs#L23) of getting `LSPEditorFeatureDetector` from services instead of from the constructor. When I do THAT (slightly different to enable testability) I don't get any hangs. We can get on a call and go into more detail if you want, I'd be happy to find that there's a less convoluted way.

Reply:
Ah, ya I recall that convo. I'm gonna dig for a little bit to see if I can find out why and if there's something we can do to resolve